server/plugin_release.go

```diff
@@ -0,0 +1,398 @@
+		return err
+	}
+
+	filepath, err := downloadAsset(assetURL)
+	if err != nil {
+		return err
+	}
+
+	fileServerPath, err := copyFileToSigningServer(filepath)
+	if err != nil {
+		return err
+	}
+
+	err = signAsset(fileServerPath)
+	if err != nil {
+		return err
+	}
+
+	filename := strings.Split(filepath, "/")
+	baseFilename := filename[len(filename)-1]
+	localSignedFiles, err := copySignedFile(baseFilename)
+	if err != nil {
+		return err
+	}
+
+	err = uploadSignedArtifcatsToGithub(ctx, client, repo, releaseID, localSignedFiles)
+	if err != nil {
+		return err
+	}
+
+	err = uploadSignedArtifcatsToS3(localSignedFiles)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkReleaseArtifact(ctx context.Context, githubClient *github.Client, repo, tag string) (releaseAssetID, releaseID int64, err error) {
+	LogInfo("Checking the if the release asset is available")
+	for {
+		repoReleases, _, err := client.Repositories.ListReleases(ctx, Cfg.GithubOrg, repo, nil)
+		if err != nil {
+			return -1, -1, err
+		}
+		for _, release := range repoReleases {
+			if release.GetTagName() == tag {
+				if release.GetAssetsURL() != "" && len(release.Assets) != 0 {
+					return release.Assets[0].GetID(), release.GetID(), nil
+				}
+				LogInfo("Release found but no assets yet. Still waiting...")
+			}
+			LogInfo("waiting...")
+		}
+
+		select {
+		case <-ctx.Done():
+			return -1, -1, errors.New("timed out waiting for ok response")
+		case <-time.After(30 * time.Second):
+		}
+	}
+}
+
+func getAssetDownloadURL(ctx context.Context, githubClient *github.Client, repo string, assetID int64) (string, error) {
+	LogInfo("Getting the asset download URL")
+	releaseAsset, _, err := client.Repositories.GetReleaseAsset(ctx, Cfg.GithubOrg, repo, assetID)
+	if err != nil {
+		return "", err
+	}
+
+	return releaseAsset.GetBrowserDownloadURL(), nil
+}
+
+func uploadSignedArtifcatsToGithub(ctx context.Context, githubClient *github.Client, repo string, releaseID int64, fileToUpload []string) error {
+	LogInfo("Uploading signed assets to Github release")
+
+	for _, file := range fileToUpload {
+		opts := &github.UploadOptions{
+			Name: file,
+		}
+
+		filePath := fmt.Sprintf("/tmp/%s", file)
+		fileHandler, err := os.Open(filePath)
+		defer fileHandler.Close()
+		if err != nil {
+			LogError("Error opening the file. err=" + err.Error())
+			return err
+		}
+		_, _, err = client.Repositories.UploadReleaseAsset(ctx, Cfg.GithubOrg, repo, releaseID, opts, fileHandler)
+		if err != nil {
+			LogError("Error while uploading to github. err=" + err.Error())
+			return err
+		}
+	}
+
+	LogInfo("Done upload to Github")
+	return nil
+}
+
+func downloadAsset(assetURL string) (string, error) {
+	LogInfo("Will download the github release asset")
+	url := strings.Split(assetURL, "/")
+	filename := url[len(url)-1]
+
+	resp, err := http.Get(assetURL)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// Create the file
+	filepath := fmt.Sprintf("/tmp/%s", filename)
```

Comment:
Hard-coding `/tmp` runs the risk of concurrent operations using the same filename.

Reply:
the artifact will have another name based on that tag and you cannot tag duplicate. but will change that to generate a temporary folder

Reply:
Sounds good -- though I note that the potential for contention might come from duplicate tags across different repositories.

Reply:
@lieut-data replace to create a temp folder PTAL